[ldc2-1.1.0-beta2] Missing symbol with inlining enabled

Compiling the following module `dt.d`

    import std.datetime;

    void main()
    {
        auto currentTime = Clock.currTime();
        auto timeString = currentTime.toISOExtString();
        auto restoredTime = SysTime.fromISOExtString(timeString);
    }

with `ldc2 --enable-inlining dt.d` results in the linker error

    dt.obj : error LNK2019: unresolved external symbol _D3std8datetime4Date14__invariant234MxFNaNfZv referenced in function _D3std8datetime4Date26__T16fromISOExtStringTAywZ16fromISOExtStringFNaNfxAywZS3std8datetime4Date
    dt.obj : error LNK2019: unresolved external symbol _D3std8datetime9TimeOfDay14__invariant265MxFNaNfZv referenced in function _D3std8datetime9TimeOfDay26__T16fromISOExtStringTAywZ16fromISOExtStringFNaNfxAywZS3std8datetime9TimeOfDay
    dt.exe : fatal error LNK1120: 2 unresolved externals

The error was reported in the forum: http://forum.dlang.org/post/tsziqjllwzthbnxmpmna@forum.dlang.org

--- a/gen/function-inlining.cpp
+++ b/gen/function-inlining.cpp
@@ -123,6 +123,17 @@ bool defineAsExternallyAvailable(FuncDeclaration &fdecl) {
     return false;
   }
 
+  // Because the frontend names `__invariant*` functions differently depending
+  // on the compilation order, we cannot emit the `__invariant` wrapper that
+  // calls the `__invariant*` functions.
+  // This is a workaround, the frontend needs to be changed such that the
+  // __invariant* names no longer depend on semantic analysis order.
+  // See https://github.com/ldc-developers/ldc/issues/1678
+  if (fdecl.isInvariantDeclaration()) {
+    IF_LOG Logger::println("__invariant cannot be emitted.");
+    return false;
+  }
+
   // TODO: Fix inlining functions from object.d. Currently errors because of
   // TypeInfo type-mismatch issue (TypeInfo classes get special treatment by the
   // compiler). To start working on it: comment-out this check and druntime will

--- a/tests/codegen/inlining_invariants_gh1678.d
+++ b/tests/codegen/inlining_invariants_gh1678.d
@@ -1,0 +1,21 @@
+// RUN: %ldc --enable-inlining -of=%t%exe %s
+
+// https://github.com/ldc-developers/ldc/issues/1678
+
+import std.datetime;
+
+// Extra test that fail when a simple frontend change is tried that names __invariant using the line and column number.
+class A {
+    mixin(genInv("666")); mixin(genInv("777"));
+}
+
+string genInv(string a) {
+    return "invariant() { }";
+}
+
+void main()
+{
+    auto currentTime = Clock.currTime();
+    auto timeString = currentTime.toISOExtString();
+    auto restoredTime = SysTime.fromISOExtString(timeString);
+}

